accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4, object? o5, object? o6, object? o7)
+        public ArgumentValuesFixed(object? obj1, object? o2, object? o3, object? o4, object? o5, object? o6, object? o7, object? o8)
+    }
+}
+```
+
+## TypedReference
+This is currently optional and being discussed. If `TypedReference` ends up supporting references to byref-like types like `Span<T>` then it will be much more useful otherwise just the existing `ref <T>` API can be used instead. The advantage of `TypedReference` is that it does not require generics so it can be made to work with `Span<T>` easier than adding a feature that would allowing generic parameters to be a byref-like type.
+
+To avoid the use of C#-only "undocumented" keywords, wrappers for `__makeref`, `__reftype`, `__refvalue` which also enable other languages.
+```diff
+namespace System
+{
+    public ref struct TypedReference
+    {
+        // Equivalent of __makeref except for a byref-like type since they can't be a generic parameter - see
+        // see https://github.com/dotnet/runtime/issues/65112 for reference.
++       public static TypedReference Make<T>(ref T? value);
++       public static unsafe TypedReference Make(Type type, void* value);
+        // Helper used for boxed or loosely-typed cases
++       public static TypedReference Make(ref object value, Type type);
+
+        // Equivalent of __refvalue
++       public ref T GetValue<T>();
+
+        // Equivalent of __reftype
++       public Type Type { get; };
+    }
+}
+```
+
+## PropertyInfo \ FieldInfo
+For `PropertyInfo`, this is an alternative of using more heavy-weight `MethodInvoker`. For `FieldInfo`, this expands on the existing `Set\GetValueDirect` to also use `TypedReference` for the `value`.
+
+```diff
+namespace System.Reflection
+{
+    public abstract class PropertyInfo
+    {
++       [System.CLSCompliantAttribute(false)]
+        public virtual void GetValueDirect(TypedReference obj, TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
+        public virtual void SetValueDirect(TypedReference obj, TypedReference value);
+
+        // Possible for performance in System.Text.Json:
++       public virtual Func<object, TValue> CreateGetterDelegate<TValue>();
++       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();
+    }
+
+    public abstract class FieldInfo
+    {
++       [System.CLSCompliantAttribute(false)]
+        public virtual void GetValueDirect(TypedReference obj, TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
+        public virtual void SetValueDirect(TypedReference obj, TypedReference value);
+
+        // Possible for performance in System.Text.Json:
++       public virtual Func<object, TValue> CreateGetterDelegate<TValue>();
++       public virtual Action<object, TValue> CreateSetterDelegate<TValue>();
```

Comment:
Perhaps `CreateGetValueDelegate` and `CreateSetValueDelegate`, to match `GetValue`, `GetValueDirect`, etc.

Reply:
(dup of above)